### ISSUE TITLE
Functions & Variables Highlighting

### ISSUE THREAD

**Issue by Kozaky** [JustusAdam/language-haskell#192]
Hi,

I recently began to learn Haskell and there is something that has made my learning a little bit harder from the very beginning. Opposite to most languages, functions are not called with parenthesis in this language and, sadly, the highlighter does not differentiate between functions and arguments, at least not in any of the color schemes I have tried.

Is there any reason for this? In my opinion, being able to differentiate between functions and arguments would  greatly improve the readability.

Thanks for your attention :)

**Comment by JustusAdam** [comment]
Hello Kozaky,

I am glad you are trying to pick up haskell, it really is a great language to learn, specifically because it makes you think differently about concepts such as functions.

The answer to your question is that in Haskell many things are actually functions and so you would just have to almost color everything again.

What do I mean by that, well in Haskell not only do you not use parentheses, but functions actually always have only one argument. Functions that seem to take multiple arguments actually only take one and the **return a new function** that also only takes one argument. 

As an example suppose we have a function `foo` that takes three arguments, one `String`, one `Int` and one `Bool` and returns `()` then the signature would be `foo :: String -> Int -> Bool -> ()` and we could apply it thus `foo "Str" 0 True`.

However we could also only apply these functions arguments one by one, yielding new functions which are called "partially applied". These can also be let bound.

```hs
let foo1 = foo "Str"
    foo2 = foo0 0
    fooResult = foo1 True
in fooResult == foo "Str" 0 True && fooResult /= foo2 False
```

As you can see all of these partial applications of the function `foo` are themselves functions again that are applied. A full application of `foo` could be parentesized as `(((foo "Str") 0) True) == foo "Str" 0 True`. Now tell me, what is the function in this case? All of the applications? Only `foo`? What in the `let` bound case, are `foo1` and `foo2` functions?

Basically this partial application stuff means it is not so easy to figure out what the "true" function to highlight is, and so in Haskell we usually don't try to do it. Hope this has helped a bit. 😉
